src/sci/lang.cljc

```diff
@@ -1,5 +1,45 @@
 (ns sci.lang
-  {:no-doc true})
+  {:no-doc true}
+  (:require [sci.impl.types]
+            [clojure.string :as str]))
 
 ;; marker interface for vars, clj only for now
 #?(:clj (definterface IVar))
+
+(defn- class-name [s]
+  (if-let [i (str/last-index-of s ".")]
+    (subs s (inc i))
+    s))
+
+(defn- package-name [s]
+  (if-let [i (str/last-index-of s ".")]
+    (subs s 0 i)
+    s))
+
+(deftype SciType [^:volatile-mutable
+                  #_:clj-kondo/ignore
+                  __data]
+  sci.impl.types.IBox
+  (getVal [_] __data)
+  (setVal [_ v] (set! __data v))
+  Object
+  (toString [_]
+    (str (:sci.impl/type-name __data)))
+  #?@(:clj
```

Comment:
TODO: get rid of IMeta / IObj implementation, these were added because of compatibility

We need to have a way to create and update the type in a way that is accessible by SCI, but not by the SCI user (types should be considered opaque by them)